### PR TITLE
webhook gitbook

### PR DESCRIPTION
**Pull request by mengqiy**

### CODE REVIEW

File: docs/book/beyond_basics/sample_webhook.md [hunk 1]
```diff
@@ -0,0 +1,207 @@
+# Webhook Example
+
+This chapter walks through a simple webhook implementation.
+
+It uses the [controller-runtime](https://godoc.org/sigs.k8s.io/controller-runtime/pkg) libraries to implement
+the Webhook and Manager. The webhook can run with other controllers in the same manager.
+
+## Setup
+
+#### Way to Deploy your Webhook Server
```
Comment (pwittrock) [inline docs/book/beyond_basics/sample_webhook.md]: Introduce what the Webhook Server is and why it is needed in the first paragraph.

Comment (pwittrock) [reply docs/book/beyond_basics/sample_webhook.md]: Actually in the concept doc :P

File: docs/book/beyond_basics/sample_webhook.md [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+1. Where the serving certificates live.
+1. In what environment the webhook server runs, in a pod or directly on a VM, etc.
+1. If in a pod, on what type of node, worker nodes or master node.
+
+The recommended way to deploy the webhook server is
+
+1. Run the webhook server as a regular pod on worker nodes through a workload API, e.g. Deployment or StatefulSet.
+1. Put the certificate in a k8s secret in the same namespace as the webhook server
+1. Mount the secret as a volume in the pod
+1. Create a k8s service to front the webhook server.
```
Comment (pwittrock) [inline docs/book/beyond_basics/sample_webhook.md]: Hopefully the config for this is generated for you by the scaffolding tools?

File: docs/book/beyond_basics/sample_webhook.md [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+`Name("yourname")` and `Path("/yourpath")` respectively.
+
+{% sample lang="go" %}
+```go
+wh, err := builder.NewWebhookBuilder().
+    Mutating().
+    Operations(admissionregistrationv1beta1.Create).
+    ForType(&corev1.Pod{}).
+    Handlers(&podAnnotator{}).
+	WithManager(mgr).
```
Comment (pwittrock) [inline docs/book/beyond_basics/sample_webhook.md]: Is the indenting for this off?  WithManager is indented further.

File: docs/book/beyond_basics/what_is_a_webhook.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+Webhooks are HTTP callbacks, providing a way for notifications to be delivered to an external web server.
+A web application implementing webhooks will send an HTTP request (typically POST) to other application when certain event happens. 
+In kubernetes world, there are 3 kinds of webhooks: admission webhook, authorization webhook and CRD conversion webhook.
+
+## Admission Webhook
+
+Admission webhooks are HTTP callbacks that receive admission requests, process them and return admission responses.
+There are two types of admission webhooks: mutating admission webhook and validating admission webhook.
+With mutating admission webhooks, you may change the requests to enforce custom policy, e.g. injecting a sidecar container
```
Comment (pwittrock) [inline docs/book/beyond_basics/what_is_a_webhook.md]: `With mutating admission webhooks, you may change the request object before it is stored (e.g. for performing defaulting of fields or enforcing a custom policy)

File: docs/book/beyond_basics/what_is_a_webhook.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+With validating admission webhooks, you may reject the requests to enforce custom admission policies, e.g. image whitelisting.
```
Comment (pwittrock) [inline docs/book/beyond_basics/what_is_a_webhook.md]: `With validating admission webhooks, you may not change the request, but you can reject it (e.g. for implementing validation of the request).`

File: docs/book/beyond_basics/what_is_a_webhook.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+#### Why Admission Webhook is Important
```
Comment (pwittrock) [inline docs/book/beyond_basics/what_is_a_webhook.md]: nit: Why Admission Webhooks are Important

File: docs/book/beyond_basics/what_is_a_webhook.md [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+Admission webhook is the mechanism to enable kubernetes extensibility through CRD.
```
Comment (pwittrock) [inline docs/book/beyond_basics/what_is_a_webhook.md]: s/is/are/

File: docs/book/beyond_basics/what_is_a_webhook.md [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+- Mutating admission webhook is the only way to do defaulting for CRDs.
+- Validating admission webhook is the only to do CRD validation that is not covered by per field basis OpenAPI validation.
```
Comment (pwittrock) [inline docs/book/beyond_basics/what_is_a_webhook.md]: Validating admission webhook allows for more complex validation than pure Schema validation.

File: docs/book/beyond_basics/what_is_a_webhook.md [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+It can also be used to add custom logic in the core kubernetes API.
+
+#### Mutating Admission Webhook
+
+A mutating admission webhook receives an admission request which contains an object.
+The webhook can either decline the request directly or returning JSON patches for modifying the original object.
+- If admitting the request, the webhook is responsible for generating JSON patches and send them back in the 
+admission response.
+- If declining the request, a reason message should be returned in the admission response.
```
Comment (pwittrock) [inline docs/book/beyond_basics/what_is_a_webhook.md]: what http code does this return to the user?

File: docs/book/beyond_basics/what_is_a_webhook.md [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+#### Validating Admission Webhook
+
+A validating admission webhook receives an admission request which contains an object.
+The webhook can either admit or decline the request.
+A reason message should be returned in the admission response if declining the request.
+
+#### Authentication
+
+The apiserver by default doesn't authenticate itself to the webhooks.
```
Comment (pwittrock) [inline docs/book/beyond_basics/what_is_a_webhook.md]: Is this the auth for the requesting user, or the apiserver itself?

File: docs/book/beyond_basics/sample_webhook.md [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+
+The recommended way to deploy the webhook server is
+
+1. Run the webhook server as a regular pod on worker nodes through a workload API, e.g. Deployment or StatefulSet.
+1. Put the certificate in a k8s secret in the same namespace as the webhook server
+1. Mount the secret as a volume in the pod
+1. Create a k8s service to front the webhook server.
+
+#### Creating a Webhook Handler
+
```
Comment (pwittrock) [inline docs/book/beyond_basics/sample_webhook.md]: Introduce the Webhook Handler.

`
The business logic for a Webhook written by implementing the `admission.Handler` interface.  This interface contains a single `Handle` method that is invoked for each request the Webhook is configured to accept.
`

File: docs/book/beyond_basics/sample_webhook.md [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+var _ admission.Handler = &podAnnotator{}
+
+func (a *podAnnotator) Handle(ctx context.Context, req types.Request) types.Response {
+	...
+}
+```
+{% endmethod %}
+
+
+#### Building a Webhook with the Handler
```
Comment (pwittrock) [inline docs/book/beyond_basics/sample_webhook.md]: I would call this:

> ## Configuring a Webhook and Registering the Handler

File: docs/book/beyond_basics/sample_webhook.md [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+1. If in a pod, on what type of node, worker nodes or master node.
+
+The recommended way to deploy the webhook server is
+
+1. Run the webhook server as a regular pod on worker nodes through a workload API, e.g. Deployment or StatefulSet.
+1. Put the certificate in a k8s secret in the same namespace as the webhook server
+1. Mount the secret as a volume in the pod
+1. Create a k8s service to front the webhook server.
+
+#### Creating a Webhook Handler
```
Comment (pwittrock) [inline docs/book/beyond_basics/sample_webhook.md]: I would call this:

> #### Creating a Handler

File: docs/book/beyond_basics/sample_webhook.md [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+	WithManager(mgr).
+	Build()
+if err != nil {
+	// handle error
+}
+```
+{% endmethod %}
+
+
+#### Creating a Webhook Server
```
Comment (pwittrock) [inline docs/book/beyond_basics/sample_webhook.md]: > ### Creating a Server and Registering the Webhook

File: docs/book/beyond_basics/sample_webhook.md [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+}
+```
+{% endmethod %}
+
+
+#### Building a Webhook with the Handler
+
+{% method %}
+
+[controller-runtime](https://godoc.org/sigs.k8s.io/controller-runtime/pkg) provides a useful package for
```
Comment (pwittrock) [inline docs/book/beyond_basics/sample_webhook.md]: Introduce the Webhook.  e.g.

`
A Webhook configures what type of requests the Handler should accept from the apiserver.  Options include:
- The type of the Operations (CRUD)
- The type of the Targets (Deployment, Pod, etc)
- The type of the Handler (Mutating, Validating)

When the server starts, the Webhook will register itself with the apiserver to start accepting requests.
`

File: docs/book/beyond_basics/sample_webhook.md [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+}
+```
+{% endmethod %}
+
+
+#### Creating a Webhook Server
+
+{% method %}
+
+You can create a new webhook server by invoking `webhook.NewServer`.
```
Comment (pwittrock) [inline docs/book/beyond_basics/sample_webhook.md]:
```
A Server registers Webhooks with the apiserver and creates an HTTP server to route requests to the handlers.  The server is behind by a Kubernetes Service and provides a certificate to the apiserver when serving requests.
```

File: docs/book/beyond_basics/sample_webhook.md [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+You can specify `Port`, `CertDir` and various `BootstrapOptions`.
+It is strongly recommended to set the `Secret` and `Service` fields.
+In other words, it is recommended to use a k8s secret to store the certificates and
+a k8s service to front this webhook server.
+
+{% sample lang="go" %}
+```go
+svr, err := webhook.NewServer("foo-admission-server", mgr, webhook.ServerOptions{
```
Comment (pwittrock) [inline docs/book/beyond_basics/sample_webhook.md]: should this be `_, err :=`

File: docs/book/beyond_basics/sample_webhook.md [hunk 18]
```diff
@@ -0,0 +1,207 @@
...
+    Port:    9876,
```
Comment (pwittrock) [inline docs/book/beyond_basics/sample_webhook.md]: Is there a good default for this?

File: docs/book/beyond_basics/sample_webhook.md [hunk 19]
```diff
@@ -0,0 +1,207 @@
...
+
+#### Adding a Webhook to the Webhook Server
+
+{% method %}
+
+You can register webhook(s) in the webhook server by invoking `Register`.
+
+{% sample lang="go" %}
+```go
+err = svr.Register(wh)
```
Comment (pwittrock) [inline docs/book/beyond_basics/sample_webhook.md]: nit: you can make this one line

File: docs/book/beyond_basics/sample_webhook.md [hunk 20]
```diff
@@ -0,0 +1,207 @@
...
+        },
+    },
+})
+if err != nil {
+    // handle error
+}
+```
+{% endmethod %}
+
+#### Adding a Webhook to the Webhook Server
```
Comment (pwittrock) [inline docs/book/beyond_basics/sample_webhook.md]: ### Registering a Webhook with the Server

File: docs/book/beyond_basics/sample_webhook.md [hunk 21]
```diff
@@ -0,0 +1,207 @@
...
+
+{% method %}
+
+`podAnnotator` is an admission webhook handler. It implements the `admission.Handler` interface.
+Details about how to implement an admission webhook podAnnotator is covered in a later section.
+
+{% sample lang="go" %}
+```go
+type podAnnotator struct {
+	client  client.Client
```
Comment (pwittrock) [inline docs/book/beyond_basics/sample_webhook.md]: Add a comment that these are automatically populated by the Webhook

File: docs/book/beyond_basics/sample_webhook.md [hunk 22]
```diff
@@ -0,0 +1,207 @@
...
+	if err != nil {
+		return admission.ErrorResponse(http.StatusBadRequest, err)
+	}
+	copy := pod.DeepCopy()
+
+	err = a.mutatePodsFn(ctx, copy)
+	if err != nil {
+		return admission.ErrorResponse(http.StatusInternalServerError, err)
+	}
+	return admission.PatchResponse(pod, copy)
```
Comment (pwittrock) [inline docs/book/beyond_basics/sample_webhook.md]: This is cool.  Add a comment about what this does.

File: docs/book/beyond_basics/sample_webhook.md [hunk 23]
```diff
@@ -0,0 +1,207 @@
...
+## Implementing Webhook Handler
+
+
+#### Implementing the Handler Business Logic
+
+{% method %}
+
+`decoder types.Decoder` is a decoder that knows how the decode all core type and your CRD types.
+
+`client client.Client` is a client that can read other objects from the manager's cache and write directly to the API server.
```
Comment (pwittrock) [inline docs/book/beyond_basics/sample_webhook.md]: Should it use the cache?  Maybe it should do live lookups?

File: docs/book/beyond_basics/sample_webhook.md [hunk 24]
```diff
@@ -0,0 +1,207 @@
...
+}
+```
+{% endmethod %}
+
+
+#### Injecting the dependencies
+
+{% method %}
+
+`podAnnotator` implements `inject.Client` and `inject.Decoder` interfaces.
```
Comment (pwittrock) [inline docs/book/beyond_basics/sample_webhook.md]: I'd move this under the Handler section.

File: docs/book/beyond_basics/sample_webhook.md [hunk 25]
```diff
@@ -0,0 +1,207 @@
...
+func (a *podAnnotator) InjectClient(c client.Client) error {
+a.client = c
+return nil
+}
+
+var _ inject.Decoder = &podAnnotator{}
+
+// InjectDecoder injects the decoder into the podAnnotator
+func (a *podAnnotator) InjectDecoder(d types.Decoder) error {
+a.decoder = d
```
Comment (pwittrock) [inline docs/book/beyond_basics/sample_webhook.md]: indenting is off

File: docs/book/beyond_basics/what_is_a_webhook.md (hunk 10 above)
Comment (mengqiy) [reply docs/book/beyond_basics/what_is_a_webhook.md]: The apiserver itself

File: docs/book/beyond_basics/what_is_a_webhook.md (hunk 9 above)
Comment (mengqiy) [reply docs/book/beyond_basics/what_is_a_webhook.md]: IMO we can document these details in the sample webhook.

File: docs/book/beyond_basics/sample_webhook.md (hunk 2 above)
Comment (mengqiy) [reply docs/book/beyond_basics/sample_webhook.md]: The service is created by the webhook server during bootstrapping.
scaffolding tools doesn't need to worry about the service.

**Comment by mengqiy** [comment]
@pwittrock PTAL

File: docs/book/beyond_basics/what_is_a_webhook.md [hunk 26]
```diff
@@ -0,0 +1,81 @@
+# Webhook
+
+Webhooks are HTTP callbacks, providing a way for notifications to be delivered to an external web server.
+A web application implementing webhooks will send an HTTP request (typically POST) to other application when certain event happens. 
+In the kubernetes world, there are 3 kinds of webhooks: admission webhook, authorization webhook and CRD conversion webhook.
```
[the remark this answers is not in the document]
Comment (Liujingfang1) [reply docs/book/beyond_basics/what_is_a_webhook.md]: Can you add comments somewhere that we only support mutating and validating webhooks?

File: docs/book/beyond_basics/sample_webhook.md [hunk 27]
```diff
@@ -0,0 +1,228 @@
...
+
+{% method %}
+
+A Server registers Webhook Configuration with the apiserver and creates an HTTP server to route requests to the handlers.
+
+The server is behind a Kubernetes Service and provides a certificate to the apiserver when serving requests.
+
+The Server depends on a Kubernetes Secret containing this certificate to be mounted under `CertDir`.
+
+The Secret doesn't have to be pre-populated, the server can populate the secret with certificate.
```
Comment (pwittrock) [inline docs/book/beyond_basics/sample_webhook.md]: secret with a generated certificate if the Secret does not exist.

File: docs/book/beyond_basics/sample_webhook.md [hunk 28]
```diff
@@ -0,0 +1,228 @@
...
+
+A new webhook server can be created by invoking `webhook.NewServer`.
+The Server will be registered to the provided manager.
+You can specify `Port`, `CertDir` and various `BootstrapOptions`.
+
+It is strongly recommended to set the `Secret` and `Service` fields.
```
Comment (pwittrock) [inline docs/book/beyond_basics/sample_webhook.md]: Delete this paragraph, it raises more questions than answers.

Comment (pwittrock) [reply docs/book/beyond_basics/sample_webhook.md]: For the full list of options Server options see <link to godoc>.

File: docs/book/beyond_basics/sample_webhook.md [hunk 29]
```diff
@@ -0,0 +1,228 @@
...
+
+#### Implementing the Handler Business Logic
+
+{% method %}
+
+`decoder types.Decoder` is a decoder that knows how the decode all core type and your CRD types.
+
+`client client.Client` is a client that knows how to talk to the API server.
+
+Note: If using a caching client, you don't need explicitly tell the cache what resources to watch.
```
Comment (pwittrock) [inline docs/book/beyond_basics/sample_webhook.md]: Note: the client.Client provided to the InjectClient function reads from a cache which is lazily initialized.  To eagerly initialize the cache, perform a read operation with the client before starting the server.

File: docs/book/beyond_basics/sample_webhook.md [hunk 30]
```diff
@@ -0,0 +1,228 @@
...
+The cache will automatically watch the resources after you do a LIST or GET on that type of resources. 
+
+The guideline of returning HTTP status code is that:
+- If the server decides to admit the request, it should return 200.
+- If the request is not well formatted, the server should reject it with 400 (Bad Request).
+- If the server encounters an unexpected error during processing, it should reject the request with 500 (Internal Error).
+- If the server rejects the request due to an admission policy reason, it should return 200.
```
[the remark this answers is not in the document]
Comment (pwittrock) [reply docs/book/beyond_basics/sample_webhook.md]: move the last one to the second position

**Comment by pwittrock** [comment]
minor nits: lgtm

**Comment by mengqiy** [comment]
Addressed all comments.
This should be good to go.

**Pull request merged**